sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala

```diff
@@ -969,4 +969,15 @@ class ParquetPartitionDiscoverySuite extends QueryTest with ParquetTest with Sha
       ))
     }
   }
+
+  test("SPARK-18108 Partition discovery fails with explicitly written long partitions") {
+    withTempPath { dir =>
+      val path = dir.getCanonicalPath
+
+      val df = Seq((1L, 2.0)).toDF("a", "b")
+      // partition explicitly written
+      df.write.parquet(s"$path/a=1")
```

Comment:
@maropu I think you're doing the wrong thing. You're having the value `a=1` as part of your data schema and your partition directory, which is not allowed. I would say the correct behavior here would be to fail the reading case, because the value is both in the partition schema and the data schema, and it's possible that they may not be equal.
For example:
```scala
val df = Seq((1L, 2.0)).toDF("a", "b")
df.write.parquet(s"$path/a=2")
// this should fail because `a` is both part of the partitioning and schema
checkAnswer(spark.read.parquet(s"$path"), Seq(Row(1L, 2.0)))
```